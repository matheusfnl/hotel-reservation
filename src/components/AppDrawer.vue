<template>
  <Transition name="fade">
    <div class="background" v-if="drawer" @click="drawer = false" />
  </Transition>

  <q-drawer
    v-model="drawer"
    :width="900"
    behavior="desktop"
    side="right"
    overlay
    bordered
    class="drawer-custom"
    @hide="() => setOverflow('auto')"
  >
    <component :is="component" v-bind="props" />
  </q-drawer>
</template>

<script lang="ts" setup>
import { ref, inject } from 'vue'
import type { Emitter } from 'mitt'

import type { DrawerEventType } from '@/types'

const emitter = inject<Emitter<DrawerEventType>>('emitter')
const drawer = ref(false)
const component = ref<object | null>({})
const props = ref<object | null>({})

emitter?.on('drawer-open', ({ component: drawerComponent, props: drawerProps }) => {
  drawer.value = true
  component.value = drawerComponent
  props.value = drawerProps
  setOverflow('hidden')
})

emitter?.on('drawer-close', () => {
  drawer.value = false
  component.value = {}
  props.value = {}
  setOverflow('auto')
})

const setOverflow = (value: string) => {
  document.body.style.overflow = value
}
</script>

<style scoped>
.background {
  width: 100vw;
  height: 100vh;
  background: black;
  position: fixed;
  z-index: 10;
  opacity: 0.3;
  top: 0;
}

.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.3s ease;
}

.fade-enter-from,
.fade-leave-to {
  opacity: 0;
}
</style>
