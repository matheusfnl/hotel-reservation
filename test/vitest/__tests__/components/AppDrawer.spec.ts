import { describe, it, expect, vi } from 'vitest'
import { mount } from '@vue/test-utils'
import { nextTick, defineComponent } from 'vue'
import mitt from 'mitt'

import AppDrawer from '@/components/AppDrawer.vue'

const emitter = mitt()

const MockComponent = defineComponent({
  template: '<div class="mock-component">Mock Component</div>',
})

const mountWrapper = () => {
  return mount(AppDrawer, {
    global: {
      provide: {
        emitter,
      },
    },
  })
}

vi.useFakeTimers()

describe('AppDrawer', () => {
  it('should render drawer with default state', () => {
    const wrapper = mountWrapper()

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.mock-component').exists()).toBe(false)
  })

  it('should opens drawer on drawer-open event', async () => {
    const wrapper = mountWrapper()

    emitter.emit('drawer-open', { component: MockComponent, props: {} })
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.mock-component').exists()).toBe(true)
  })

  it('should closes drawer on drawer-close event', async () => {
    const wrapper = mountWrapper()

    emitter.emit('drawer-open', { component: MockComponent, props: {} })
    await nextTick()

    emitter.emit('drawer-close')
    await nextTick()

    vi.advanceTimersByTime(300)
    await nextTick()

    expect(wrapper.html()).toMatchSnapshot()
    expect(wrapper.find('.mock-component').exists()).toBe(false)
  })
})
