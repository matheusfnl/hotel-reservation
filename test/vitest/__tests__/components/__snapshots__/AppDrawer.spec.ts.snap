// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AppDrawer > should closes drawer on drawer-close event 1`] = `
"<transition-stub data-v-99ee5784="" name="fade" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>
<div data-v-99ee5784="" class="q-drawer-container">
  <aside class="q-drawer q-drawer--right q-drawer--bordered q-layout--prevent-focus q-drawer--standard fixed q-drawer--on-top" style="width: 900px; transform: translateX(900px);">
    <div class="q-drawer__content fit scroll drawer-custom">
      <!---->
    </div>
  </aside>
</div>"
`;

exports[`AppDrawer > should opens drawer on drawer-open event 1`] = `
"<transition-stub data-v-99ee5784="" name="fade" appear="false" persisted="false" css="true">
  <div data-v-99ee5784="" class="background"></div>
</transition-stub>
<div data-v-99ee5784="" class="q-drawer-container">
  <aside class="q-drawer q-drawer--right q-drawer--bordered q-drawer--standard fixed q-drawer--on-top" style="width: 900px; transform: translateX(0px);">
    <div class="q-drawer__content fit scroll drawer-custom">
      <div data-v-99ee5784="" class="mock-component">Mock Component</div>
    </div>
  </aside>
</div>"
`;

exports[`AppDrawer > should render drawer with default state 1`] = `
"<transition-stub data-v-99ee5784="" name="fade" appear="false" persisted="false" css="true">
  <!--v-if-->
</transition-stub>
<div data-v-99ee5784="" class="q-drawer-container">
  <aside class="q-drawer q-drawer--right q-drawer--bordered q-layout--prevent-focus q-drawer--standard fixed q-drawer--on-top" style="width: 900px; transform: translateX(900px);">
    <div class="q-drawer__content fit scroll drawer-custom">
      <!---->
    </div>
  </aside>
</div>"
`;
